tf2onnx/onnx_opset/math.py

```diff
@@ -555,6 +557,207 @@ def version_11(cls, ctx, node, **kwargs):
+        graph.add_graph_input(utils.make_name("condition_in"), onnx_pb.TensorProto.BOOL, [])
+        graph.add_graph_input(utils.make_name("inputs"), inputs_tensor_dtype, inputs_tensor_shape)
+        graph.add_graph_input(utils.make_name("axis_length_plus_one"), onnx_pb.TensorProto.INT64, [1])
+        graph.add_graph_input(utils.make_name("accumulator"), inputs_tensor_dtype, inputs_tensor_shape)
+        if not isinstance(axis, int): # axis is a tensor, we need to feed it to the loop graph
+            graph.add_graph_input(utils.make_name("axis"), onnx_pb.TensorProto.INT64, [1])
+            axis = graph.input_names[-1]
+            axis_node = graph.get_node_by_output(axis)
+        else:
+            axis_node = graph.make_const(utils.make_name("axis"), np.array([axis], "int64"))
+
+        # main loop graph
+        loop_name = node.name + "/loop"
+        iter_num = GraphBuilder(graph).make_unsqueeze({'data': graph.input_names[0], 'axes': [0]})
+        one_node = graph.make_const(utils.make_name("one"), np.array(1, "int64"))
+        zero_node = graph.make_const(utils.make_name("zero"), np.array([0], "int64"))
+
+        add_node = graph.make_node("Add", inputs=[iter_num, one_node.output[0]],
+                                   outputs=[utils.make_name("add")], op_name_scope=loop_name)
+
+        if node.get_attr_value("reverse"):
+            pad_axis = [zero_node.output[0], add_node.output[0]]
+            start_slice = add_node.output[0]
+            end_slice = graph.input_names[3]
+        else:
+            neg_node = graph.make_node("Neg", inputs=[add_node.output[0]],
+                                       outputs=[utils.make_name("neg")], op_name_scope=loop_name)
+            pad_axis = [add_node.output[0], zero_node.output[0]]
+            start_slice = zero_node.output[0]
+            end_slice = neg_node.output[0]
+
+        pads_node = cls.get_pads_node(graph, pad_axis, axis, input_rank, is_pad_axis_const=False, base_name=loop_name)
+        slice_node = graph.make_node("Slice", op_name_scope=loop_name, outputs=[utils.make_name("slice")],
+                                     inputs=[graph.input_names[2], start_slice, end_slice, axis_node.output[0]])
+        if graph.get_dtype(slice_node.output[0]) != graph.get_dtype(one_node.output[0]):
+            pad_const_node = graph.make_node("Cast", inputs=[one_node.output[0]],
+                                             attr={"to": graph.get_dtype(slice_node.output[0])},
+                                             op_name_scope=loop_name, outputs=[utils.make_name("pad_const")])
+        else:
+            pad_const_node = one_node
+        pad_node = graph.make_node("Pad", inputs=[slice_node.output[0], pads_node.output[0], pad_const_node.output[0]],
+                                   op_name_scope=loop_name, outputs=[utils.make_name("pad")])
+        mul_node = graph.make_node("Mul", inputs=[graph.input_names[4], pad_node.output[0]],
```

Comment:
Is it possible to leverage reduce_prod to avoid multiplying those values one by one?

Reply:
sure, didn't think about that. I will refactor the code in order to use that operator

Reply:
@fatcat-z Actually, using ReduceProd, would require to loop over slices of the input tensor to be reduced and then concatenate results to create the output tensor. However, in order to support a non-constant axis for the CumProd op as in TF, I don't think it is possible to use the Concat op, since it requires the axis as an attribute. Due to this limitation, I don't see how I can use ReduceProd instead of Mul.